src/viur/core/bones/uri.py

```diff
@@ -0,0 +1,204 @@
+                ])
+
+            elif "-" in accepted_ports:  # range of ports
+                start, end = accepted_ports.split("-", 1)
+                start = int(start)
+                end = int(end)
+                if start > end:
+                    raise ValueError("Start value must be less than end value")
+
+                if start < PORT_MIN:
+                    raise ValueError("Start value must be greater than zero")
+
+                if end > PORT_MAX:
+                    raise ValueError(f"End value must be less or equal than {PORT_MAX}")
+
+                return [range(start, end + 1)]
+
+            else:
+                port = int(accepted_ports)
+                return [range(port, port + 1)]
+
+        elif isinstance(accepted_ports, int):
+            if accepted_ports < PORT_MIN:
+                raise ValueError("Port value must be greater than zero")
+
+            if accepted_ports > PORT_MAX:
+                raise ValueError(f"Port value must be less or equal than {PORT_MAX}")
+
+            return [range(accepted_ports, accepted_ports + 1)]
+
+        elif isinstance(accepted_ports, Iterable):
+            accepted_ports_value = []
+            for accepted_port in accepted_ports:
+                accepted_ports_value.extend(UriBone._build_accepted_ports(accepted_port))
+            return accepted_ports_value
+
+        raise ValueError("accepted_ports must be a iterable or an integer or string")
+
+    def isInvalid(self, value) -> str | None:
+        try:
+            parsed_url = urlparse(value)
+        except ValueError:
+            return "Can't parse URL"
+
+        if not self.local_path_allowed and parsed_url.scheme == "":
+            return f"""No protocol specified"""
+
+        if self.accepted_ports:
+            if not any(parsed_url.port in rng for rng in self.accepted_ports):
+                return f""""{parsed_url.port}" not in the accepted ports."""
+
+        if self.accepted_protocols:
+            for protocol in self.accepted_protocols:
+                if fnmatch.fnmatch(parsed_url.scheme, protocol):
+                    break
+            else:
+                return f""""{parsed_url.scheme}" not in the accepted protocols."""
+
+        if self.domain_allowed_list is not None:
+            for domain in self.domain_allowed_list:
+                if fnmatch.fnmatch(value, domain):
+                    break
+            else:
+                return f"""Url is not in the domain allowed list."""
+
+        if self.domain_disallowed_list is not None:
+            for domain in self.domain_disallowed_list:
+                if fnmatch.fnmatch(value, domain):
+                    return f"""Url is in the domain disallowed list."""
+
+    def singleValueFromClient(self, value, skel, bone_name, client_data) -> tuple:
+        if err := self.isInvalid(value):
+            return self.getEmptyValue(), [ReadFromClientError(ReadFromClientErrorSeverity.Invalid, err)]
```

Comment:
Return `value` here, otherwise anything the user inputs is simply emptied, which is not user-friendly.
```suggestion
            return value, [ReadFromClientError(ReadFromClientErrorSeverity.Invalid, err)]
```